src/disk_io_thread.cpp

```diff
@@ -2792,12 +2792,20 @@ namespace libtorrent
 		cached_piece_entry* pe = m_disk_cache.find_piece(j);
 		if (pe == nullptr) return 0;
 
+		piece_refcount_holder refcount_holder(pe);
+
 #if TORRENT_USE_ASSERTS
 		pe->piece_log.push_back(piece_log_t(j->action));
 #endif
 		try_flush_hashed(pe, m_settings.get_int(
```

Comment:
sometimes the problem comes from `do_flush_hashed` and sometimes (much less) from `do_flush_piece`. My question (more than a theory) is: why both methods don't handle the call to `try_flush_hashed` in the same way?

Maybe this is the missing "memory pin" that prevents the segfault, but frankly, I don't have a clear idea on this.